os/pm/pm_metrics.c

```diff
@@ -230,14 +251,68 @@ void pm_metrics_update_changestate(void)
  *   None
  *
  ****************************************************************************/
-void pm_metrics_update_wakehandler(clock_t missing_tick, pm_wakeup_reason_code_t wakeup_src)
+void pm_metrics_update_wakehandler(uint32_t missing_tick, pm_wakeup_reason_code_t wakeup_src)
 {
 	if (g_pm_metrics_running) {
+		g_pm_metrics->thread_metrics.wakeup_src = wakeup_src;
+		g_pm_metrics->thread_metrics.delay = missing_tick;
 		g_pm_metrics->wakeup_src_counts[wakeup_src]++;
 		g_pm_metrics->board_sleep_ticks += missing_tick;
 	}
 }
 
+/****************************************************************************
+ * Name: pm_metrics_update_sleep
+ *
+ * Description:
+ *   This function is called inside pm_sleep's callback. It counts the frequency of
+ *   thread which wakeup the board. It also checks the minimum amount of time board
+ *   was in sleep because of given thread.
+ * 
+ * Input parameters:
+ *   pid - the ID of thread
+ *
+ * Returned value:
+ *   None
+ *
+ ****************************************************************************/
+void pm_metrics_update_sleep(pid_t pid)
+{
+	int hash_pid;
+	if (g_pm_metrics_running && (g_pm_metrics->thread_metrics.wakeup_src == PM_WAKEUP_HW_TIMER)) {
+		hash_pid = PIDHASH(pid);
+		if (g_pm_metrics->thread_metrics.delay < g_pm_metrics->thread_metrics.min_sleep_ticks[hash_pid]) {
+			g_pm_metrics->thread_metrics.min_sleep_ticks[hash_pid] = g_pm_metrics->thread_metrics.delay;
+		}
+		g_pm_metrics->thread_metrics.wakeup_counts[hash_pid]++;
+		g_pm_metrics->thread_metrics.wakeup_src = PM_WAKEUP_UNKNOWN;
+	}
+}
+
+/****************************************************************************
+ * Name: pm_metrics_update_recover
+ *
+ * Description:
+ *   This function is called inside pm_recover. It resets the wakeup_counts and 
+ *   sleep_ticks of given thread for consistent PM Metrics result.
+ * 
+ * Input parameters:
+ *   pid - the ID of thread
+ *
+ * Returned value:
+ *   None
+ *
+ ****************************************************************************/
+void pm_metrics_update_recover(pid_t pid)
+{
+	int hash_pid;
+	if (g_pm_metrics_running) {
+		hash_pid = PIDHASH(pid);
+		g_pm_metrics->thread_metrics.min_sleep_ticks[hash_pid] = UINT32_MAX;
```

Comment:
We can modify this data when the next pid comes in.
For this, do not add action to task end.